### ISSUE TITLE
django-todo & wagtail integration

### ISSUE THREAD

**Issue by Fandekasp** [shacker/django-todo#69]
Hi,

I'm evaluating what are my options to add a task & writer management system to the [Wagtail CMS](https://docs.wagtail.io/en/v2.4/index.html), and I discovered django-todo through [this comment on ls.joyous](https://github.com/linuxsoftware/ls.joyous/issues/7#issuecomment-485625149).

I will spend more time studying the internals of django-todo to see if I could start a wagtail-todo project (other names I have in mind are wagtail-tasks and wagtail-workflow).

First of all, I cannot see django-todo's roadmap. If there is one, can you point it out? It looks like you've been maintaining it for a while, which is reassuring, but that would be upsetting if you were to stop maintaining it tomorrow :)

Some features that I would need, but could be added directly to django-todo, are:
* Recurring tasks (time-based, trigger action based)
* Templates (managed from django-admin, ManyToManyField with Groups, select dropdown list in form)

Some unique features I can think of adding in a project reconcialing wagtail and django-todo are:
* wagtail-admin styled templates (I suppose I'll take inspiration from your [GTD](https://github.com/shacker/gtd) app)
* hooks to auto-complete and auto-create tasks (latter if recurring task feature doesn't exist in django-todo) based on some event, and default templates to figure out what task to complete/create depending on the model action being performed (sounds shaky, but trying to link tasks with some contenttype, some field and some field value opens a hole of complexity)

What are your thoughts on this? I'll be keen on contributing to django-todo core, but I have a tight schedule (I'll need to have something shipped in production by the end of next month), so your support will be highly appreciated :)

**Comment by shacker** [comment]
@Fandekasp I'm a huge fan of wagtail, and used it on a work project for a couple of years, but am no longer current with wagtail. That said, I'd love to see an integration! 

First, I think you should separate out from your list above the things that have nothing to do with wagail. Please feel free to create *separate* issues for:

- Recurring tasks
- Trigger-action tasks (these are two quite different things as I see it)

If I recall, the way to get Django models to appear in the Wagtail admin is to have them inherit from the Wagtail Page class. That unfortunately is *not* something we want to do in this core project, since it needs to work in any site. If there are other ways to get Django models to appear, you should look into that and write your own plugin or bridge module that you can distribute (I'll link to it). That would cover your wagtail admin templates too. 

So basically, happy to take contributions for scheduled and triggered tasks, but wagtail-specific functionality should exist as an outside module. The last thing we need is a bunch of conditional code to support the specifics of a bunch of different CMSs.

**Comment by Fandekasp** [comment]
@shacker thanks for the quick reply. Glad to see we're thinking alike.

I only recently discovered Wagtail, but I understand we can either [extend ModelAdmin](http://docs.wagtail.io/en/v2.4/reference/contrib/modeladmin/index.html) or [register snippets](https://docs.wagtail.io/en/v2.4/topics/snippets.html?highlight=snippet). That's something I would handle in a wagtail-todo app.

I will create separate issues for each component, but would you mind telling me more about django-todo's plans for the future? Is there no roadmap? Will you at least keep django-todo up to date with the test Django version?

**Comment by shacker** [comment]
The only roadmap is the features you see mentioned in the Issues here, and the priority is whatever I decide to work on next, or what someone contributes code for :)

And yes, this project is going on 11 years now. It's evolved a ton in that time! It's open source software - if I stopped working on it, I would hand off responsibility to someone else, or you would have 142 forks to choose from :)

**Comment by Fandekasp** [comment]
alright, fair enough :) 

I'll spend some time learning about django-todo and see how it could fit with wagtail, and if I decide to go with it, I'll create the new issues and we can go from there. Thanks again for the fast support!

**Comment by datatalking** [comment]
@Fandekasp were you able to make any progress on using this with wagtail?
